From b7992270c264f04bbb48825e8a0ae84330491fee Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Karel=20Ko=C4=8D=C3=AD?= <karel.koci@nic.cz>
Date: Wed, 17 Apr 2019 10:53:59 +0200
Subject: [PATCH] ubus: force restart in case ubus is not able to communicate
 with ubusd

When there is a huge update and ubus client utility is no longer able to
communicate with running ubusd we have to restart ubusd to run new
updated version. We are sure that we are running updated version after
restart because it is a dependency of ubus it self so it is installed
before it. We are not able to just reload it because ubus is run by
procd and we are not able to reload pid 1. Fortunately procd restarts
ubusd when it crashes. We can abuse that and restart it by killing it.
This effectively reloads it.
We don't want to do it on every update becuase killing ubus might have
some other effect we are not aware of. We do it only when we fail to
communicate with ubusd.

We should also consider what happens if procd is unable to communicate
with ubus. That would meant that we are unable to restart services and
such. The problem is that only interface procd currently exports is up
to my knowledge trough ubus. That means that if user-space is unable to
communicate with ubusd because both ubus utility and library were
updated and are unable to communicate with it then it has same effect
like procd not able to communicate with ubusd. The advantage of
reloading ubusd is to allow rest of the system to communicate.

The real reason why we want this is to make update faster. Any call to
ubus timeouts and that takes around a minute. This is every service
restart, every configuration update and so on. This means that from just
few minutes long update this makes it few hours long update.
---
 package/system/ubus/Makefile | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/package/system/ubus/Makefile b/package/system/ubus/Makefile
index ef492d2c65..4e44c8183f 100644
--- a/package/system/ubus/Makefile
+++ b/package/system/ubus/Makefile
@@ -72,6 +72,13 @@ define Package/libubus-lua/install
 	$(CP) $(PKG_BUILD_DIR)/lua/ubus.so $(1)/usr/lib/lua/
 endef
 
+define Package/ubus/postinst
+[ -n "$${IPKG_INSTROOT}" ] || {
+# Force restart of ubusd when ubus utility is no longer able to communicate with it
+ubus call system info 2>/dev/null >&2 || killall ubusd
+}
+endef
+
 $(eval $(call BuildPackage,libubus))
 $(eval $(call BuildPackage,libubus-lua))
 $(eval $(call BuildPackage,ubus))
-- 
2.21.0

